### PR TITLE
Implement Nonhydrostatic2DModel

### LINKED ISSUE #91: Implement a Nonhydrostatic2DModel module
**context**
Now that we have hybrid domain support, we should be able to set up a non-hydrostatic 2D model on 1+1D domains.

**ideas**
Replicate the module setup from the ShallowWaterModels and the SingleColumnModels.

**next steps**
Set up the constructors and port over the equation set for a Cartesian box (for now we stay flat) and set up the relevant test cases.

### PR DESCRIPTION
**Pull request by bischtob**
closes #91. 

This PR added the nonhydrostatic 2D model for the hybrid plane, and added a unit test for the rising bubble benchmark.

### CODE REVIEW

**Comment by bischtob** [comment]
This is ready for review.

**Comment by bischtob** [comment]
bors try

File: src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        # potential temperature
+        @. dρθ = -hdiv(ρuh * ρθ / ρ)
+        @. dρθ -= vector_vdiv_f2c(ρw * scalar_interp_c2f(ρθ / ρ))
+
+        # horizontal momentum
+        @. dρuh = -hdiv(ρuh ⊗ ρuh / ρ + pressure(ρθ) * I)
+        @. dρuh -= tensor_vdiv_f2c(ρw ⊗ vector_interp_c2f(ρuh / ρ))
+
+        # vertical momentum
+        # TODO: How ugly is this?! (-_-)' . ￣□￣｜｜  (ಥ﹏ಥ)
```
Comment (szy21) [inline src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl:119]: ¯_(ツ)_/¯

File: src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        uh_f = @. vector_interp_c2f(ρuh / ρ)
+        @. dρw = hdiv(uh_f ⊗ ρw)
```
Comment (szy21) [inline src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl:121]: Comparing with [here](https://github.com/CliMA/ClimaCore.jl/blob/1a3c9463d367aa4dede9a11ffd81249ea8fa9017/examples/hybrid/bubble.jl#L198), should this be `@. dρw = -hdiv(uh_f ⊗ ρw)`?

File: src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        ρuh = Ym.ρuh
+        ρw = Ym.ρw
+        ρθ = Ym.ρθ
+
+        # advection, pressure gradients, sources
+        # density
+        @. dρ = -hdiv(ρuh)
+        @. dρ -= vector_vdiv_f2c(ρw)
+
+        # potential temperature
```
Comment (szy21) [inline src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl:110]:
```suggestion
        # potential temperature density
```

File: test/test_cases/run_dry_rising_bubble_2d.jl [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    FT;
+    stepper = SSPRK33(),
+    nelements = (10, 50),
+    npolynomial = 4,
+    dt = 0.02,
+    callbacks = (),
+    mode = :regression,
+)
+    if FT <: Float32
+        @info "Ekman column 1D test does not run for $FT."
```
Comment (szy21) [inline test/test_cases/run_dry_rising_bubble_2d.jl:13]:
```suggestion
        @info "Dry rising bubble 2D test does not run for $FT."
```

**Comment by szy21** [comment]
Looks great! I left a few comments. I saw the link to #110 has been removed - is it because the regression and validation tests haven't been set up?

File: src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl (hunk 1 above)
Comment (jakebolewski) [reply src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl:119]: ideally you would have a preset field for computing this so you dont' have to allocate every rhs evaluation (for now).

File: src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl (hunk 2 above)
Comment (jakebolewski) [reply src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl:121]: yes

File: src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+        # vertical momentum
+        # TODO: How ugly is this?! (-_-)' . ￣□￣｜｜  (ಥ﹏ಥ)
+        uh_f = @. vector_interp_c2f(ρuh / ρ)
+        @. dρw = hdiv(uh_f ⊗ ρw)
+        @. dρw += B(
+            Geometry.transform(
+                Geometry.Cartesian3Axis(),
+                -(scalar_grad_c2f(pressure(ρθ))) +
+                scalar_interp_c2f(ρ) * Geometry.Covariant3Vector(-g), # TODO!: Not generally a Covariant3Vector
```
Comment (jakebolewski) [inline src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl:126]: you could use `grad(\Phi(z))` here to get the right output

File: src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+            ) - tensor_vdiv_c2f(tensor_interp_f2c(
+                ρw ⊗ ρw / scalar_interp_c2f(ρ),
+            )),
+        )
+
+        # TODO! diffusion goes here!
+
+        # discrete stiffness summation for spectral operations
+        Spaces.weighted_dss!(dρ)
```
Comment (jakebolewski) [inline src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl:135]: it's probably going to be a lot more efficient to use fields of multiple variables here

File: src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl (hunk 2 above)
Comment (jiahe23) [reply src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl:121]: Good catch! Thanks!

File: src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl (hunk 6 above)
Comment (jiahe23) [reply src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl:135]: @jakebolewski  
Agreed!
What would be the correct syntax? Does something like the following work?
```
Spaces.weighted_dss!(dYm)
```

File: src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl (hunk 5 above)
Comment (jiahe23) [reply src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl:126]: @jakebolewski  I guess you meant something like this line [here](https://github.com/CliMA/ClimaCore.jl/blob/1a3c9463d367aa4dede9a11ffd81249ea8fa9017/examples/hybrid/bubble.jl#L194).
It is a bit hard to access the coordinates and this is only a hacky way of doing it for now...
Do you think if there's a simple way to access `coords.z` in the way the model is currently written?

**Comment by jiahe23** [comment]
> Looks great! I left a few comments. I saw the link to #110 has been removed - is it because the regression and validation tests haven't been set up?

Yep, this PR only add RTB to unit test making sure it does bot break. Let's add the rest in a separate PR ;)

File: src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl (hunk 5 above)
Comment (jakebolewski) [reply src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl:126]: yes, it is fine for now as a hacky soln but I think eventually you'll want to be able to grab references to the coordnate data (I assume mostly z) so it is worth thinking about how best to pass that information through.  Accessing it as a field is just a passed reference to a view so it's really cheap.

File: src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl (hunk 6 above)
Comment (jakebolewski) [reply src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl:135]: Yes, ideally you would aggregate all variables on their respective spaces (center, faces).  The DSS will do the right thing for all the variables on  that particular center, face field.

File: src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl (hunk 5 above)
Comment (jiahe23) [reply src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl:126]: Sounds good!

File: src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl (hunk 6 above)
Comment (jiahe23) [reply src/Models/Nonhydrostatic2DModels/nonhydrostatic_2d_model.jl:135]: Sounds good. This also propagates to the shallow water model. I open a new issue #124 and will deal with it separately.

**Review by jiahe23 — APPROVED** [review]

**Pull request merged**